xaitk_saliency/impls/gen_object_detector_blackbox_sal/occlusion_based.py

```diff
@@ -0,0 +1,172 @@
+
+        pert_dets = blackbox.detect_objects(
+            occlude_image_streaming(
+                ref_image,
+                pert_masks,
+                fill=self.fill,
+                threads=self._threads
+            )
+        )
+
+        pert_dets_mat = _dets_to_formatted_mat(pert_dets)
+
+        return self._generator(
+            ref_dets_mat,
+            pert_dets_mat,
+            pert_masks
+        )
+
+    @classmethod
+    def get_default_config(cls) -> Dict[str, Any]:
+        cfg = super().get_default_config()
+        cfg['perturber'] = make_default_config(PerturbImage.get_impls())
+        cfg['generator'] = make_default_config(GenerateDetectorProposalSaliency.get_impls())
+        return cfg
+
+    @classmethod
+    def from_config(
+        cls: Type[C],
+        config_dict: Dict,
+        merge_default: bool = True
+    ) -> C:
+        config_dict = dict(config_dict)  # shallow-copy
+        config_dict['perturber'] = from_config_dict(
+            config_dict['perturber'],
+            PerturbImage.get_impls()
+        )
+        config_dict['generator'] = from_config_dict(
+            config_dict['generator'],
+            GenerateDetectorProposalSaliency.get_impls()
+        )
+        return super().from_config(config_dict, merge_default=merge_default)
+
+    def get_config(self) -> Dict[str, Any]:
+        return {
+            "perturber": to_config_dict(self._perturber),
+            "generator": to_config_dict(self._generator),
+            "threads": self._threads,
+        }
+
+
+def _dets_to_formatted_mat(
+    dets: Iterable[Iterable[Tuple[AxisAlignedBoundingBox, Dict[Hashable, float]]]]
+) -> np.ndarray:
+    """
+    Converts detections, as returned by an implementation of
+    ``DetectImageObjects``, into a detection matrix formatted for use with
+    an implementation of ``GenerateDetectorProposalSaliency``. The order of the
+    class scores is the same as is returned by the ``DetectImageObjects``
+    implementation.
+
+    :param dets: Detections, as returned by an implementation of
+        ``DetectImageObjects``.
+
+    :returns: Matrix of detections with shape
+        [nImgs x nDets x (4+1+nClasses)]. If the number of detections for
+        each image is not consistent, the matrix will be padded with rows
+        of ones, except for the objectness which is set to zero.
+    """
+
+    dets_mat_list = []
+    for img_dets in dets:
+
+        img_dets = list(img_dets)
+
+        if len(img_dets) == 0:
+            dets_mat_list.append(np.empty([0, 1]))
+            continue
+
+        bbox_mat = np.asarray([
+            [
+                bbox.min_vertex[0],
+                bbox.min_vertex[1],
+                bbox.max_vertex[0],
+                bbox.max_vertex[1]
+            ] for bbox in [det[0] for det in img_dets]
+        ])
+
+        score_dicts = [det[1] for det in img_dets]
+        scores_mat = np.array([list(d.values()) for d in score_dicts])
```

Comment:
Relying on dictionary values iterable ordering seems a little dangerous/fragile to me. Maybe an input to this method could be a class label list to drive the scores extracted / order here?

Reply:
Added a class label input to do this. It grabs the labels from the first perturbed detection result it finds and uses those for the rest.